src/Multipart/AbstractUploader.php

```diff
@@ -0,0 +1,455 @@
+<?php
+namespace Aws\Multipart;
+
+use Aws\AwsClientInterface as Client;
+use Aws\CommandInterface;
+use Aws\CommandPool;
+use Aws\Exception\AwsException;
+use Aws\Exception\MultipartUploadException;
+use Aws\Result;
+use Aws\ResultInterface;
+use GuzzleHttp\Promise;
+use GuzzleHttp\Promise\PromiseInterface;
+use GuzzleHttp\Psr7;
+use InvalidArgumentException as IAE;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\StreamableInterface as Stream;
+
+/**
+ * Encapsulates the execution of a multipart upload to S3 or Glacier.
+ */
+abstract class AbstractUploader
+{
+    /** @var Client Client used for the upload. */
+    protected $client;
+
+    /** @var array Configuration used to perform the upload. */
+    protected $config;
+
+    /** @var array Service-specific information about the upload workflow. */
+    protected $info = [];
+
+    /** @var PromiseInterface Promise that represents the multipart upload. */
+    protected $promise;
+
+    /** @var Stream Source of the data to be uploaded. */
+    protected $source;
+
+    /** @var UploadState State used to manage the upload. */
+    protected $state;
```

Comment:
I use `$this->state->{methodOnUploadState}` everywhere though, so it needs to be an `UploadState`. Maybe I should track the promise that resolves to the `UploadState` in a different variable?